local_ads/statistics.cpp

```diff
@@ -0,0 +1,526 @@
+auto constexpr kDeletionPeriod = std::chrono::hours(24);
+
+// TODO: set correct address
+std::string const kStatisticsServer = "";
+
+void WriteMetadata(FileWriter & writer, std::string const & countryId, int64_t mwmVersion,
+                   local_ads::Timestamp const & ts)
+{
+  local_ads::WriteCountryName(writer, countryId);
+  local_ads::WriteZigZag(writer, mwmVersion);
+  local_ads::WriteTimestamp<std::chrono::seconds>(writer, ts);
+}
+
+void ReadMetadata(ReaderSource<FileReader> & src, std::string & countryId, int64_t & mwmVersion,
+                  local_ads::Timestamp & ts)
+{
+  countryId = local_ads::ReadCountryName(src);
+  mwmVersion = local_ads::ReadZigZag(src);
+  ts = local_ads::ReadTimestamp<std::chrono::seconds>(src);
+}
+
+void WritePackedData(FileWriter & writer, local_ads::Statistics::PackedData && packedData)
+{
+  WriteToSink(writer, packedData.m_eventType);
+  WriteToSink(writer, packedData.m_zoomLevel);
+  WriteToSink(writer, packedData.m_featureIndex);
+  WriteToSink(writer, packedData.m_seconds);
+  local_ads::WriteZigZag(writer, packedData.m_mercator);
+  WriteToSink(writer, packedData.m_accuracy);
+}
+
+using ReadCallback = std::function<void(local_ads::Statistics::PackedData &&, std::string const &,
+                                        int64_t, local_ads::Timestamp const &)>;
+
+void ReadPackedData(ReaderSource<FileReader> & src, ReadCallback const & callback)
+{
+  if (callback == nullptr)
+    return;
+
+  using PackedData = local_ads::Statistics::PackedData;
+
+  std::string countryId;
+  int64_t mwmVersion;
+  local_ads::Timestamp baseTimestamp;
+  ReadMetadata(src, countryId, mwmVersion, baseTimestamp);
+  while (src.Size() > 0)
+  {
+    PackedData data;
+    data.m_eventType = ReadPrimitiveFromSource<uint8_t>(src);
+    data.m_zoomLevel = ReadPrimitiveFromSource<uint8_t>(src);
+    data.m_featureIndex = ReadPrimitiveFromSource<uint32_t>(src);
+    data.m_seconds = ReadPrimitiveFromSource<uint32_t>(src);
+    data.m_mercator = local_ads::ReadZigZag(src);
+    data.m_accuracy = ReadPrimitiveFromSource<uint16_t>(src);
+    callback(std::move(data), countryId, mwmVersion, baseTimestamp);
+  }
+}
+
+local_ads::Timestamp GetMinTimestamp(std::list<local_ads::Event> const & events,
+                                     std::string const & countryId, int64_t mwmVersion)
+{
+  local_ads::Timestamp minTimestamp = local_ads::Timestamp::max();
+  for (auto const & event : events)
+  {
+    if (event.m_countryId != countryId || event.m_mwmVersion != mwmVersion)
+      continue;
+
+    if (event.m_timestamp < minTimestamp)
+      minTimestamp = event.m_timestamp;
+  }
+  return minTimestamp;
+}
+
+local_ads::Timestamp GetMaxTimestamp(std::list<local_ads::Event> const & events,
+                                     std::string const & countryId, int64_t mwmVersion)
+{
+  local_ads::Timestamp maxTimestamp = local_ads::Timestamp::min();
+  for (auto const & event : events)
```

Comment:
This loop is the same as the loop above. How about to write a small filter:
```c++
template <typename ToDo>
void FilterEvents(std::list & events, string const & countryId, int64_t mwmVersion, ToDo && toDo) {
   ...
}
```
and use it here and above?